src/librustc_typeck/check/closure.rs

```diff
@@ -620,6 +637,104 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         result
     }
 
+    /// Invoked when we are translating the generator that results
+    /// from desugaring an `async fn`. Returns the "sugared" return
+    /// type of the `async fn` -- that is, the return type that the
+    /// user specified. The "desugared" return type is a `impl
+    /// Future<Output = T>`, so we do this by searching through the
+    /// obligations to extract the `T`.
+    fn deduce_future_output_from_obligations(
```

Comment:
```suggestion
    fn ty_infer_future_output_from_obligations(
```

Reply:
Hmm, I think I prefer the current name. For one thing, `ty_infer` is not a verb; to me, this name suggests that we return a `ty::Infer` value or something like that. For another, the `deduce_` is used throughout this file under its plain English meaning of "figure out". I think I deliberately did not use "infer" as the choice of verb there because this inference mechanism is not related to the inference variables and other techniques we typically use.